src/Xamarin.Android.Build.Tasks/Tasks/CompileNativeAssembly.cs

```diff
@@ -0,0 +1,211 @@
+		}
+
+		[Required]
+		public ITaskItem[] Sources { get; set; }
+
+		[Required]
+		public bool DebugBuild { get; set; }
+
+		[Required]
+		public string WorkingDirectory { get; set; }
+
+		public override bool Execute ()
+		{
+			try {
+				return DoExecute ();
+			} catch (Exception e) {
+				Log.LogCodedError ("XA3001", $"{e}");
+				return false;
+			}
+		}
+
+		bool DoExecute ()
+		{
+			Yield ();
+			try {
+				var task = ThreadingTasks.Task.Run ( () => RunParallelAssembler (), Token);
+				task.ContinueWith (Complete);
+
+				base.Execute ();
+
+				if (!task.Result)
+					return false;
+			} finally {
+				Reacquire ();
+			}
+
+			return !Log.HasLoggedErrors;
+		}
+
+		bool RunParallelAssembler ()
+		{
+			try {
+				ThreadingTasks.ParallelOptions options = new ThreadingTasks.ParallelOptions {
+					CancellationToken = Token,
+					TaskScheduler = ThreadingTasks.TaskScheduler.Default,
+				};
+
+				ThreadingTasks.Parallel.ForEach (GetAssemblerConfigs (), options,
+					config => {
+						if (!RunAssembler (config)) {
+							LogCodedError ("XA3001", $"Could not compile native assembly file: {Path.GetFileName (config.InputSource)}");
+							Cancel ();
+							return;
+						}
+					}
+				);
+			} catch (OperationCanceledException) {
+				return false;
+			}
+
+			return true;
+		}
+
+		bool RunAssembler (Config config)
+		{
+			var stdout_completed = new ManualResetEvent (false);
+			var stderr_completed = new ManualResetEvent (false);
+			var psi = new ProcessStartInfo () {
+				FileName = config.AssemblerPath,
+				Arguments = config.AssemblerOptions,
+				UseShellExecute = false,
+				RedirectStandardOutput = true,
+				RedirectStandardError = true,
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+				WorkingDirectory = WorkingDirectory,
+			};
+
+			string assemblerName = Path.GetFileName (config.AssemblerPath);
+			LogDebugMessage ($"[Native Assembler] {psi.FileName} {psi.Arguments}");
+			using (var proc = new Process ()) {
+				proc.OutputDataReceived += (s, e) => {
+					if (e.Data != null)
+						OnOutputData (assemblerName, s, e);
+					else
+						stdout_completed.Set ();
+				};
+
+				proc.ErrorDataReceived += (s, e) => {
+					if (e.Data != null)
+						OnErrorData (assemblerName, s, e);
+					else
+						stderr_completed.Set ();
+				};
+
+				proc.StartInfo = psi;
+				proc.Start ();
+				proc.BeginOutputReadLine ();
+				proc.BeginErrorReadLine ();
+				Token.Register (() => { try { proc.Kill (); } catch (Exception) { } });
+				proc.WaitForExit ();
+
+				if (psi.RedirectStandardError)
+					stderr_completed.WaitOne (TimeSpan.FromSeconds (30));
+
+				if (psi.RedirectStandardOutput)
+					stdout_completed.WaitOne (TimeSpan.FromSeconds (30));
+
+				return proc.ExitCode == 0;
+			}
+		}
+
+		IEnumerable<Config> GetAssemblerConfigs ()
+		{
+			string sdkBinDirectory = MonoAndroidHelper.GetOSBinPath ();
+			foreach (ITaskItem item in Sources) {
+				string abi = item.GetMetadata ("abi");
```

Comment:
Should this call `.ToLowerInvariant()` so the `abi` is "known"?

I haven't yet tracked where `Sources` comes from, but if it's user available, the casing could differ.

Reply:
It comes from `Xamarin.Android.Common.targets` and is item metadata, potentially user can modify it (not sure what for, but it's possible)